<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

Схема данных для работы над проектом приведена [здесь](https://github.com/esmelnikova/education_dst_128/tree/main/Project_2/description/Схема_данных.md)

In [252]:
import pandas as pd
import psycopg2

In [253]:
# параметры подключения к БД
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = '***' # введите пароль
HOST = '84.201.134.129'
PORT = 5432

In [254]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [255]:
# текст запроса
query_3_1 = f'''select count(*) from public.vacancies'''

In [256]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\3515521326.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [257]:
query_3_2 = f'''select count(*) from public.employers'''

In [258]:
df = pd.read_sql_query(query_3_2, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\438664466.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [259]:
query_3_3 = f'''select count(*) from public.areas'''

In [260]:
df = pd.read_sql_query(query_3_3, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\2888627660.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [261]:
query_3_4 = f'''select count(*) from public.industries'''

In [262]:
df = pd.read_sql_query(query_3_4, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\181467837.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

**Выводы по предварительному анализу данных**

Схема данных содержит 5 основных таблиц для анализа вакансий:
* public.vacancies (вакансии)
* public.employers (работодатели)
* public.areas (регионы)
* public.industries (сферы деятельности предприятий)
* public.employers_industries (для связи "много-ко-многим" между таблицами employers и industries)
    
Таблицы не пустые, данные во вакансиям присутсвуют в достаточном количестве (около 50 тыс. вакансий у 23 тыс. работодателей), 
возможен переход к более детальному анализу данных.

Справочники регионов и сфер деятельности содержат множество неповторяющихся значений (соответсвенно 1392 и 294), что делает эти признаки слабоинформативным для анализа. Рекомендуется в дальнейшем преобразование данных с объединением признаков в группы.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [263]:
query_4_1 = f'''
    SELECT a.name as area, count(v.id) as cnt
    FROM public.vacancies v
    JOIN public.areas a
    ON v.area_id=a.id
    GROUP BY a.id
    ORDER BY cnt DESC
'''

In [264]:
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\2004364207.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [265]:
query_4_2 = f'''
    SELECT count(*) 
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [266]:
df = pd.read_sql_query(query_4_2, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\1373709454.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [267]:
query_4_3 = f'''
    SELECT round(avg(salary_from)) as avg_salary_from,
	       round(avg(salary_to)) as avg_salary_to
    FROM public.vacancies
'''

In [268]:
df = pd.read_sql_query(query_4_3, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\511848361.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

In [269]:
# минимальные и максимальные значения для границ зарплатной вилки
query_4_3_1 = f'''
    SELECT  min(salary_from) as min_salary_from, max(salary_from) as max_salary_from,
	        min(salary_to) as min_salary_to, max(salary_to) as max_salary_to
    FROM public.vacancies
'''

df = pd.read_sql_query(query_4_3_1, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\491911997.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3_1, connection)


min_salary_from  max_salary_from  min_salary_to  max_salary_to
0               30          1000000             60        1000000

In [270]:
# минимальные и максимальные значения для границ зарплатной вилки в зависимости от региона
query_4_3_2 = f'''
    SELECT  a. name,
	        min(salary_from) as min_salary_from, max(salary_from) as max_salary_from,
	        min(salary_to) as min_salary_to, max(salary_to) as max_salary_to
    FROM public.vacancies v
	JOIN public.areas a ON v.area_id=a.id
	GROUP BY a.id
	ORDER BY min_salary_from, max_salary_from DESC, min_salary_to, max_salary_to DESC
'''

df = pd.read_sql_query(query_4_3_2, connection)
display(df.head(50))

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\253061076.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3_2, connection)


name  min_salary_from  max_salary_from  \
0                        Сочи             30.0         500000.0   
1                   Челябинск             30.0         300000.0   
2                     Воронеж             35.0         270000.0   
3                    Смоленск             60.0         180000.0   
4                  Нур-Султан             65.0         400000.0   
5                      Самара             70.0         300000.0   
6                         Уфа             80.0         400000.0   
7                       Химки             80.0         250000.0   
8                 Новосибирск            105.0         500000.0   
9                      Москва            250.0         650000.0   
10                  Хабаровск            750.0         250000.0   
11                    Тбилиси            900.0         350691.0   
12                      Брест            900.0         140254.0   
13                      Минск           1000.0         350691.0   
14                  Махачкала           1000.0         200000.0   
15                Владикавказ           1000.0         200000.0   
16  Киров (Кировская область)           1000.0         180000.0   
17               Новороссийск           1000.0         120000.0   
18                   Кемерово           1500.0         200000.0   
19                  Сыктывкар           1500.0         100000.0   
20            Нижний Новгород           2000.0         355000.0   
21                     Гродно           2805.0         110000.0   
22                    Ташкент           4506.0         292242.0   
23                Калининград           5000.0         355000.0   
24                Владивосток           5000.0         280000.0   
25                       Чита           5000.0         150000.0   
26            Санкт-Петербург           6000.0        1000000.0   
27                      Тосно           6000.0          95100.0   
28                         Ош           7282.0           7282.0   
29                     Алматы           7323.0         574000.0   
30                      Актау           7323.0          83585.0   
31              Петропавловск           7323.0          36614.0   
32                     Актобе           7323.0          36614.0   
33                  Кызылорда           7323.0          30512.0   
34                 Ставрополь           8000.0         200000.0   
35                    Саранск           8000.0         180000.0   
36                   Костанай           8543.0          85433.0   
37                Талдыкорган           8543.0          36614.0   
38                    Могилев           9350.0          81828.0   
39                  Солигорск           9350.0          35064.0   
40                   Мурманск           9600.0         178000.0   
41                  Караганда           9764.0          73228.0   
42                  Жезказган           9764.0          40276.0   
43                   Шахтинск           9764.0           9764.0   
44                     Казань          10000.0         500000.0   
45                 Красноярск          10000.0         500000.0   
46                Севастополь          10000.0         355000.0   
47               Екатеринбург          10000.0         355000.0   
48                       Омск          10000.0         350000.0   
49                  Краснодар          10000.0         350000.0   

    min_salary_to  max_salary_to  
0         25000.0      1000000.0  
1          6000.0       350000.0  
2         25000.0       500000.0  
3            60.0       300000.0  
4         12205.0       526036.0  
5         20000.0       467588.0  
6         25000.0       800000.0  
7         50000.0       300000.0  
8         24000.0       584485.0  
9           350.0      1000000.0  
10          850.0       350000.0  
11        58448.0       500000.0  
12        28051.0       200000.0  
13         2338.0       526036.0  
14        25000.0       300000.0  
15        29000.0       250000.0  
16         2

In [271]:
# минимальные и максимальные значения для границ зарплатной вилки в зависимости от сферы деятельности
query_4_3_2 = f'''
    SELECT  ind.name,
	        min(salary_from) as min_salary_from, max(salary_from) as max_salary_from,
	        min(salary_to) as min_salary_to, max(salary_to) as max_salary_to
    FROM public.vacancies v
	JOIN public.employers emp ON v.employer_id=emp.id
	JOIN public.employers_industries emp_ind ON emp.id=emp_ind.employer_id
	JOIN public.industries ind ON emp_ind.industry_id=ind.id
	GROUP BY ind.id
	ORDER BY min_salary_from, max_salary_from DESC, min_salary_to, max_salary_to DESC
'''

df = pd.read_sql_query(query_4_3_2, connection)
display(df.head(50))

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\1221558397.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3_2, connection)


name  min_salary_from  \
0                              Строительство жилищное             30.0   
1                     Агентские услуги в недвижимости             30.0   
2                 Безопасность, охранная деятельность             30.0   
3                         Государственные организации             35.0   
4                 Разработка программного обеспечения             42.0   
5   Интернет-компания (поисковики, платежные систе...             60.0   
6   Системная интеграция,  автоматизации технологи...             65.0   
7   Оборудование для ресторанов, гостиниц, кейтери...             70.0   
8   Оборудование для ресторанов, гостиниц, кейтери...             70.0   
9                               Консалтинговые услуги             80.0   
10                  Колл-центры, диспетчерские службы             80.0   
11                                               Банк           1000.0   
12                                   Складские услуги           1000.0   
13                            Автомобильные перевозки           1000.0   
14                       Страхование, перестрахование           1500.0   
15  Сельскохозяйственная продукция (продвижение, о...           2922.0   
16  Маркетинговые, рекламные, BTL, дизайнерские, E...           3000.0   
17                        Обучение иностранным языкам           3506.0   
18                          Издательская деятельность           5143.0   
19                               Тренинговые компании           5143.0   
20      Розничная сеть (электроника, бытовая техника)           5301.0   
21  Услуги по ведению бухгалтерского и налогового ...           5301.0   
22                Лечебно-профилактические учреждения           6000.0   
23                        Архитектура, проектирование           6102.0   
24                                 Кадровые агентства           7013.0   
25                                   Интернет-магазин           7282.0   
26                                 Интернет-провайдер           7323.0   
27                                     Аптека, оптика           8000.0   
28                                 Центры развлечения           8000.0   
29                                 Школа, детский сад           8000.0   
30                             Вуз, ссуз колледж, ПТУ           8543.0   
31                      Курьерская, почтовая доставка           9350.0   
32                           Юридическое обслуживание           9600.0   
33                                    Мобильная связь           9764.0   
34                               Оптоволоконная связь           9764.0   
35                                Фиксированная связь           9764.0   
36                      Несетевая розница, мелкий опт           9764.0   
37  Строительство коммерческих объектов (торговые ...          10000.0   
38                  Консалтинг, оценка в недвижимости          10000.0   
39  Управляющая, инвестиционная компания (управлен...          10000.0   
40             Управление и эксплуатация недвижимости          10000.0   
41                              Мебель (производство)          10000.0   
42                            Розничная сеть (мебель)          10000.0   
43             Мебель (продвижение, оптовая торговля)          10000.0   
44  Производство мультимедиа, контента, редакторск...          10000.0   
45  Электронно-вычислительная, оптическая, контрол...          10000.0   
46  Бытовая техника, электроника, климатическое об...          10000.0   
47                                 Продюсерский центр          10000.0   
48             Общественная, политическая организация          10417.0   
49  Фармацевтическая продукция (продвижение, оптов...          10603.0   

    max_salary_from  min_salary_to  max_salary_to  
0          500000.0        21206.0       200000.0  
1          500000.0        25000.0      1000000.0  
2          160000.0        17064.0       140000.0  
3          250000.0        12205.0       300000.0  
4          500

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [272]:
query_4_4 = f'''
    SELECT format('%s - %s',schedule,employment) as type, count(id)
    FROM public.vacancies
    GROUP BY schedule, employment
    ORDER BY count DESC
'''

In [273]:
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\4268832789.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


type  count
0           Полный день - Полная занятость  35367
1      Удаленная работа - Полная занятость   7802
2         Гибкий график - Полная занятость   1593
3   Удаленная работа - Частичная занятость   1312
4        Сменный график - Полная занятость    940
5                 Полный день - Стажировка    569
6        Вахтовый метод - Полная занятость    367
7        Полный день - Частичная занятость    347
8      Гибкий график - Частичная занятость    312
9           Полный день - Проектная работа    141
10     Удаленная работа - Проектная работа    133
11              Гибкий график - Стажировка    116
12    Сменный график - Частичная занятость    101
13           Удаленная работа - Стажировка     64
14        Гибкий график - Проектная работа     18
15             Сменный график - Стажировка     12
16       Вахтовый метод - Проектная работа      2
17       Сменный график - Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [274]:
query_4_5 = f'''
    SELECT experience, count(id)
    FROM public.vacancies
    GROUP BY experience
    ORDER BY count
'''

In [275]:
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\3611053821.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Выводы по детальному анализу вакансий**

* Наибольшее количество вакансий размещены в крупных городах-миллионниках (Москва, Санкт-Петербург, Новосибирск, Екатеринбург и т.д.). Москва - абсолютный лидер в этом рейтинге, более 5 тыс. вакансий. За ним следует Санкт-Петербург, где около 2800 вакансий.
* Рассматриваются вакансии не только на территории РФ, но и в зарубежных странах (Республика Беларусь, Казахстан, Узбекистан, Литва, Сербия, Германия и др.)

* Данные о границах зарплатой вилки заполнены только в половине вакансий (примерно 24 тыс. из 49 тыс. вакансий), этот факт необходимо учесть при разработке алгоритмов рекомендаций вакансий (Например, предсказывать эти значения из аналогичных вакансий, заполнять модальным значением или задать низкий весовой коэфиициент для этого признака при отборе подходящих вакансий и т.д.)

* Разброс значений для границ зарплатной вилки слишком велик (нижняя граница: 30-1000000 руб., верхняя: 60-1000000 руб.), что делает расчет среднего значения для этих параметров не информативным. Кроме того, разброс уровня ЗП может существенно различаться в зависимости от региона и сферы деятельности предприятия.
Рекомендуется дополнительно исследовать эти параметры, например построить распределение, вычислить модальное значение, проследить корреляцию с выше указанными признаками.
* Присутствуют потенциально аномальные значения в зарплатной вилке (слишком низкие значения: 30, 70 руб. и т.д.), возможно данные ошибочно загружены не в рублях или не учтена кратность, например тыс.руб.

* В большинстве вакансий (ок. 70%) требуются сотрудники на полный рабочий день с полной занятостью, 15% вакансий рассчитаны на удаленную работу с полной занятостью.

* В большинсте вакансий (50%) требуются соискатели с опытом работы от 1 года до 3 лет, в 30% вакансий необходимый опыт составляет от 3 до 6 лет.
Но присутствует достаточно большое кол-во вакансий (15%), где готовы взять соискателя без опыта работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [276]:
query_5_1 = f'''
    SELECT e.name, count(v.id)
    FROM public.vacancies v
    JOIN public.employers e
    ON v.employer_id=e.id
    GROUP BY e.id
    ORDER BY count DESC
    LIMIT 5
'''

In [277]:
df = pd.read_sql_query(query_5_1, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\2915140023.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [278]:
query_5_2 = f'''
    SELECT a.name, count(distinct e.id) emp_count, count(distinct v.id) vac_count --считаем кол-во уникальных(!) компаний в регионе
    FROM public.areas a
    LEFT JOIN public.employers e ON a.id=e.area
    LEFT JOIN public.vacancies v ON v.area_id=a.id
    GROUP BY a.id
    ORDER BY vac_count ASC, emp_count DESC
'''

In [279]:
df = pd.read_sql_query(query_5_2, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\2644321718.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  emp_count  vac_count
0                 Россия        410          0
1              Казахстан        207          0
2     Московская область         75          0
3     Краснодарский край         19          0
4     Ростовская область         18          0
...                  ...        ...        ...
1357              Алматы        721       1892
1358         Новосибирск        573       2006
1359               Минск       1115       2112
1360     Санкт-Петербург       2217       2851
1361              Москва       5864       5333

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [280]:
query_5_3 = f'''
    SELECT e.name, count(distinct v.area_id) as cnt_areas --считаем кол-во уникальных(!) регионов для компании
    FROM public.employers e
    JOIN public.vacancies v
    ON v.employer_id=e.id
    GROUP BY e.id
    ORDER BY cnt_areas desc
'''

In [281]:
df = pd.read_sql_query(query_5_3, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\1682211407.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  cnt_areas
0                                      Яндекс        181
1                                  Ростелеком        152
2                                  Спецремонт        116
3                      Поляков Денис Иванович         88
4                                   ООО ЕФИН          71
...                                       ...        ...
14901                            НПП Авиатрон          1
14902              Центр дистанционных торгов          1
14903  Городские Телекоммуникационные Системы          1
14904                       Введенский, Отель          1
14905              Группа Компаний Автоимпорт          1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [282]:
query_5_4 = f'''
    SELECT count(e.id)
    FROM public.employers e --сфера может быть не указана, используем внешнее соединение
    LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id 
    LEFT JOIN public.industries i ON i.id=ei.industry_id
    WHERE i.id is NULL
'''

In [283]:
df = pd.read_sql_query(query_5_4, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\1196796479.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [284]:
query_5_5 = f'''
    SELECT e.name, count(i.id)
    FROM public.employers e 
    JOIN public.employers_industries ei ON e.id=ei.employer_id 
    JOIN public.industries i ON i.id=ei.industry_id
    GROUP BY e.id
    HAVING count(i.id)=4
    ORDER BY name
'''

In [285]:
df = pd.read_sql_query(query_5_5, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\1337360513.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1133            ЮРИОН      4
1134  ЮТИП Технологии      4
1135           ЯКласс      4
1136            ЯрНео      4
1137    Яшин&Партнёры      4

[1138 rows x 2 columns]

In [286]:
query_5_5_1 = f'''
    SELECT e.name, count(i.id)
    FROM public.employers e 
    JOIN public.employers_industries ei ON e.id=ei.employer_id 
    JOIN public.industries i ON i.id=ei.industry_id
    GROUP BY e.id
    HAVING count(i.id)>1
    ORDER BY name
'''

df = pd.read_sql_query(query_5_5_1, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\3544455309.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5_1, connection)


name  count
0                                    101 Интернет      4
1               144 бронетанковый ремонтный завод      2
2                                             169      5
3     16-я городская детская поликлиника г.Минска      2
4                                         179 СРЗ      2
...                                           ...    ...
8625                      Ясли-сад №374 г. Минска      2
8626                                          ЯСП      2
8627                            Ястро-Переработка      3
8628                                         ЯТСС      3
8629                                Яшин&Партнёры      4

[8630 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [129]:
query_5_6 = f'''
    SELECT count(*)
    FROM public.employers e 
    JOIN public.employers_industries ei ON e.id=ei.employer_id 
    JOIN public.industries i ON i.id=ei.industry_id
    WHERE i.name like 'Разработка программного обеспечения'
'''

In [130]:
df = pd.read_sql_query(query_5_6, connection)
display(df)

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\2317649464.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [131]:
# получение списка городов-милионников

# найти таблицы на странице по ключевому слову
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' 
city_tb = pd.read_html(url, match='город')

# получить из таблицы список названий городов-миллионников
city_ls = list(city_tb[0]['Город']) if len(city_tb)>0 else []
print(city_ls)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж[a]', 'Пермь', 'Волгоград']


In [132]:
# убрать лишние символы из названия города
city_ls[13]='Воронеж'
print(city_ls)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [141]:
# запрос

emp_name='Яндекс'
cities = "'" + "', '".join(city_ls) + "'"

query_5_7 = f'''
WITH mc_vac AS (
	SELECT mc.name, count(vac.id) as cnt 
 	FROM
		(select id, name from public.areas where name in ({cities})
		) as mc
 	JOIN public.vacancies vac ON vac.area_id=mc.id
 	JOIN public.employers emp ON emp.id=vac.employer_id
 	WHERE emp.name like '{emp_name}'
 	GROUP BY mc.name
 ) --CTE выполняется единожды
 SELECT * FROM mc_vac
 UNION ALL
 SELECT 'Total', sum(cnt) from mc_vac
 ORDER BY cnt
'''

In [142]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\sheme\AppData\Local\Temp\ipykernel_12988\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

**Выводы по анализу работодателей**

* Всего в БД хранятся данные о 23501 работодателе.

* Наибольшее количество вакансий ожидаемо зарегистрированы от имени крупных компаний таких как: Яндекс, Ростелеком, Тинькофф, Сбер и т.д. Но абсолютным лидером является Яндекс, они предлагают в 4 раза больше вакансий, чем другие крупные компании.

* Наибольшее количество работодателей и вакансий размещены в крупных городах. Например в Москве находятся почти 25% из всех работодалей, сохраненных в БД, в Санкт-Петербурге 9% и далее этот процент уменьшается для остальных регионов.

* Среди регионов с доступными вакансиями прослеживается линейная зависимость кол-ва работодателей и числа предложений. Среди регионов, в которых нет вакансий, по количеству работодателей лидируют "Россия", "Казахстан".

* Лидерами по количеству регионов, в которых компания разместила вакансии, являются Яндекс и Ростелеком.

* У трети работодателей указано более 1 сферы деятельности.

* Около 15% работодателей ищут соискателей в области разработки ПО.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса

In [ ]:
# результат запроса

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса

In [ ]:
# результат запроса

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса

In [ ]:
# результат запроса

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса

In [ ]:
# результат запроса

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса

In [ ]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса

In [ ]:
# результат запроса

***

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования